algorithms/src/std_algorithms/Kokkos_HelperPredicates.hpp

```diff
@@ -0,0 +1,115 @@
+// notice, this list of conditions and the following disclaimer in the
+// documentation and/or other materials provided with the distribution.
+//
+// 3. Neither the name of the Corporation nor the names of the
+// contributors may be used to endorse or promote products derived from
+// this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY NTESS "AS IS" AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL NTESS OR THE
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+// LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact Christian R. Trott (crtrott@sandia.gov)
+//
+// ************************************************************************
+//@HEADER
+*/
+
+#ifndef KOKKOS_STD_HELPER_PREDICATES_HPP
+#define KOKKOS_STD_HELPER_PREDICATES_HPP
+
+// naming convetion:
+// StdAlgoSomeExpressiveNameUnaryPredicate
+// StdAlgoSomeExpressiveNameBinaryPredicate
+
+namespace Kokkos {
+namespace Experimental {
+namespace Impl {
+
+// ------------------
+// UNARY PREDICATES
+// ------------------
+template <class T>
+struct StdAlgoEqualsValUnaryPredicate {
+  T m_value;
+
+  KOKKOS_INLINE_FUNCTION
+  bool operator()(const T& val) const { return val == m_value; }
+
+  KOKKOS_INLINE_FUNCTION
+  StdAlgoEqualsValUnaryPredicate(const T& _value) : m_value(_value) {}
+};
+
+template <class T>
+struct StdAlgoNotEqualsValUnaryPredicate {
+  T m_value;
+
+  KOKKOS_INLINE_FUNCTION
+  bool operator()(const T& val) const { return !(val == m_value); }
+
+  KOKKOS_INLINE_FUNCTION
+  StdAlgoNotEqualsValUnaryPredicate(const T& _value) : m_value(_value) {}
+};
+
+template <class ValueType, class PredicateType>
+struct StdAlgoNegateUnaryPredicateWrapper {
+  PredicateType m_pred;
+
+  KOKKOS_INLINE_FUNCTION
+  bool operator()(const ValueType& val) const { return !m_pred(val); }
+
+  KOKKOS_INLINE_FUNCTION
+  StdAlgoNegateUnaryPredicateWrapper(const PredicateType& pred)
+      : m_pred(pred) {}
+};
+
+// ------------------
+// BINARY PREDICATES
+// ------------------
+template <class ValueType1, class ValueType2>
+struct StdAlgoEqualBinaryPredicate {
+  KOKKOS_INLINE_FUNCTION
+  bool operator()(const ValueType1& a, const ValueType2& b) const {
+    return a == b;
+  }
+};
+
+template <class ValueType1, class ValueType2>
+struct StdAlgoLessThanBinaryPredicate {
+  KOKKOS_INLINE_FUNCTION
+  bool operator()(const ValueType1& a, const ValueType1& b) const {
```

Comment:
Should `b` be of type `ValueType2`?

Do we care that we can't technically use this one for pointers?  Unless one is using `std::less`, you cannot do a less than comparison unless the pointers point into (or one past the end) the same array or malloced space.

We could do the trick that some standard library implementations do:  if `ValueType` is a pointer, cast it to `uintptr_t` before doing the comparison.

Reply:
> Should b be of type ValueType2?

yes, thanks for finding it!

Re the other things, I will get back to it